libzfs_core/_libzfs_core.py

```diff
@@ -0,0 +1,659 @@
+# Copyright 2015 ClusterHQ. See LICENSE file for details.
+
+"""
+Python wrappers for libzfs_core interfaces.
+
+As a rule, there is a Python function for each C function.
+The signatures of the Python functions generally follow those of the
+functions, but the argument types are natural to Python.
+nvlists are wrapped as dictionaries or lists depending on their usage.
+Some parameters have default values depending on typical use for
+increased convenience.  Output parameters are not used and return values
+are directly returned.  Error conditions are signalled by exceptions
+rather than by integer error codes.
+"""
+
+import errno
+import threading
+from . import exceptions
+from . import _error_translation as xlate
+from .bindings import libzfs_core
+from ._constants import MAXNAMELEN
+from ._nvlist import nvlist_in, nvlist_out
+
+
+def lzc_create(name, is_zvol = False, props = {}):
+    '''
+    Create a ZFS filesystem or a ZFS volume ("zvol").
+
+    :param bytes name: a name of the dataset to be created.
+    :param bool is_zvol: whether to create a zvol (false by default).
+    :param props: a `dict` of ZFS dataset property name-value pairs (empty by default).
+    :type props: dict of bytes:Any
+
+    :raises FilesystemExists: if a dataset with the given name already exists.
+    :raises ParentNotFound: if a parent dataset of the requested dataset does not exist.
+    :raises PropertyInvalid: if one or more of the specified properties is invalid
+                             or has an invalid type or value.
+
+    Wraps ``int lzc_create(const char *fsname, dmu_objset_type_t type, nvlist_t *props)``.
+    '''
+    if is_zvol:
+        ds_type = _lib.DMU_OST_ZVOL
+    else:
+        ds_type = _lib.DMU_OST_ZFS
+    with nvlist_in(props) as nvlist:
+        ret = _lib.lzc_create(name, ds_type, nvlist)
+    xlate.lzc_create_xlate_error(ret, name, is_zvol, props)
+
+
+def lzc_clone(name, origin, props = {}):
+    '''
+    Clone a ZFS filesystem or a ZFS volume ("zvol") from a given snapshot.
+
+    :param bytes name: a name of the dataset to be created.
+    :param bytes origin: a name of the origin snapshot.
+    :param props: a `dict` of ZFS dataset property name-value pairs (empty by default).
+    :type props: dict of bytes:Any
+
+    :raises FilesystemExists: if a dataset with the given name already exists.
+    :raises DatasetNotFound: if either a parent dataset of the requested dataset
+                             or the origin snapshot does not exist.
+    :raises PropertyInvalid: if one or more of the specified properties is invalid
+                             or has an invalid type or value.
+
+    .. note::
+        Because of a deficiency of the underlying C interface
+        :exc:`.DatasetNotFound` can mean that either a parent filesystem of the target
```

Comment:
I've not seen this `.name` syntax used in this context before.  Sphinx seems to understand it somehow - though it generates exactly the same markup whether I build the docs with this or with a modified version that just uses the name "DatasetNotFound" without the "." (can't figure out how to literally format what I mean in a github comment, sorry).

So I don't really know what Sphinx is actually doing here.  But since it works like this, I guess that makes it good. :smile: